### PR TITLE
feat(expr): add new crate `common-expression`

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Initial commit for the new scalar expression framework. See [tracking issue](https://github.com/datafuselabs/databend/issues/6547) for more information.

- [x] Expression definition
- [x] Expression type checker
- [x] Expression evaluator

### CODE REVIEW

**Comment by andylokandy** [comment]
```
ERROR the following files don't have a valid license header: 
common/expression/tests/it/testdata/run-ast 
```

How can I bypass this check for goldenfiles?

**Comment by PsiACE** [comment]
> How can I bypass this check for goldenfiles?

edit https://github.com/datafuselabs/databend/blob/main/.licenserc.yaml

**Comment by andylokandy** [comment]
@PsiACE But why common-ast bypass the check without being included in this config?

File: common/expression/src/function.rs [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+        assert!(
+            !has_nullable,
+            "Function {} has nullable argument or output, please use register_1_arg_core instead",
+            name
+        );
+
+        let property = property.preserve_not_null(true);
+
+        self.register_1_arg_core::<NullType, NullType, _>(name, property.clone(), move |_, _| {
+            Value::Scalar(())
```
Comment (ygf11) [inline common/expression/src/function.rs:175]: Maybe we should keep column length if it is a column?

File: common/expression/src/expression.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    ColumnRef {
+        id: usize,
+        data_type: DataType,
+        property: ValueProperty,
+    },
+    // Cast {
+    //     is_try: bool,
+    //     expr: Box<Expr>,
+    //     dest_type: DataType,
+    // },
```
[the remark this answers is not in the document]
Comment (andylokandy) [reply common/expression/src/expression.rs:34]: unless we've figured out a way to make cast a normal function, we have to keep it.

File: common/expression/src/function.rs (hunk 1 above)
Comment (andylokandy) [reply common/expression/src/function.rs:175]: Is seems not necessary to keep a column length for a scalar. Is there a place we had used?

**Comment by Xuanwo** [comment]
> @PsiACE But why common-ast bypass the check without being included in this config?

Nice question, I did some research:

https://github.com/apache/skywalking-eyes/blob/9b6b981a99fb9179f1c5aa58ac6474ea1dbdd552/pkg/header/config.go#L143

`skywalking-eyes` will ignore all `*.txt` files, and `common-ast`'s golden files are suffixed by `.txt`.

File: common/expression/src/function.rs (hunk 1 above)
Comment (ygf11) [reply common/expression/src/function.rs:175]: Thanks, got it, NullType is just an inner data type.

Comment (andylokandy) [reply common/expression/src/function.rs:175]: NullType is not a inner type, but similar to `Nullable<Nothing>` in clickhouse.

**Review by sundy-li — APPROVED** [review]
LGTM, yet needs to cover the `Object` Column.

File: common/expression/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# Workspace dependencies
+common-arrow = { path = "../arrow" }
+
+# Github dependencies
+
+# Crates.io dependencies
+chrono-tz = "0.6.1"
+enum-as-inner = "0.4"
+# https://github.com/calder/rust-goldenfile/pull/7
+goldenfile = { git = "https://github.com/datafuse-extras/rust-goldenfile", rev = "16c5783" }
```
[the remark this answers is not in the document]
Comment (Xuanwo) [reply common/expression/Cargo.toml:22]: Also cc @PsiACE, we can remove `datafuse-extras/rust-goldenfile` after `goldenfile` upgraded to `1.4.3`.

File: common/expression/src/function.rs (hunk 1 above)
Comment (ygf11) [reply common/expression/src/function.rs:175]: I think it still need column length here. It is not clear in 1 arg function.

For example, we have a 2-arg-function `plus`, the inputs are `Literal::Null` and`ColumnRef(UInt8)`.

The type check will select `plus(Null, UInt8)` as the finaly function.

The result will be a scalar, so it is not correct.

File: common/expression/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (andylokandy) [reply common/expression/Cargo.toml]: I have updated all goldenfiles to 1.4.3.

**Comment by sundy-li** [comment]
flaky sqllogic test, I did not find the reason. cc @ZeaLoVe

**Comment by ZeaLoVe** [comment]
> flaky sqllogic test, I did not find the reason. cc @ZeaLoVe

9wwoCI3cuZ

Not a flaky error, this sql error code change, expect 1006 but get 1065.

**Comment by BohuTANG** [comment]
> > flaky sqllogic test, I did not find the reason. cc @ZeaLoVe
> 
> 9wwoCI3cuZ
> 
> Not a flaky error, this sql error code change, expect 1006 but get 1065.

How can we fix it, in this PR or another?

**Comment by youngsofun** [comment]
@BohuTANG  fixed  https://github.com/datafuselabs/databend/pull/6593

**Comment by Xuanwo** [comment]
@mergify update

**Pull request merged**